### PR TITLE
make Timestamp parsing consistent across Java versions

### LINKED ISSUE #19195: make Timestamp parsing consistent across Java versions
- [x] Determine where `Instant.parse` is being used and if there are any backwards compatibility issues
- [x] Identify and resolve potential issues (possibly following a small design doc?)

To be resolved for both `main` and `main-2.x`

References:
- Original JDK bug: https://bugs.openjdk.org/browse/JDK-8166138
- Quick fix for issue: https://github.com/digital-asset/daml/pull/17053
- Canton upstream failure: https://github.com/DACH-NY/canton/pull/13688
- Canton main-2.x JDK 17/11 work https://github.com/DACH-NY/canton/pull/19063

### PR DESCRIPTION
**Pull request by carlpulley-da**
Fixes #19195 

- [x] package protect `Timestamp.assertFromString` and `Timestamp.fromString`
- [x] replace non-LF call sites with a direct use of `Instant.parse`
- [x] ensure `DateTimeParseException` is managed with new uses of `Instant.parse`

### CODE REVIEW

**Review by yves-da — COMMENTED** [review]
Changes look fine to me. Thanks!

But I don't consider myself an authority in judging whether this accomplishes what is stated: That is a consistent timestamp parsing for Java 11 and Java 17.

**Comment by carlpulley-da** [comment]
> But I don't consider myself an authority in judging whether this accomplishes what is stated: That is a consistent timestamp parsing for Java 11 and Java 17.

As these changes are for `main`, my understanding is that we're only using Java 17 here. Hence these changes (i.e. use of `Instant.parse`) are sufficient here.

For `main-2.x`, we'll need to work a little harder!

File: sdk/canton/community/ledger/ledger-json-api/src/test/scala/com/digitalasset/canton/daml/lf/value/json/ApiCodecCompressedSpec.scala [hunk 1]
```diff
@@ -328,8 +328,7 @@ class ApiCodecCompressedSpecStable extends ApiCodecCompressedSpec {
         "\"1990-11-09T04:30:23.1234569Z\"",
       ),
       c("\"1970-01-01T00:00:00Z\"", VA.timestamp)(Time.Timestamp assertFromLong 0),
-      // TODO(i12203) enable the following line once we move away from Java 11 -- see https://bugs.openjdk.org/browse/JDK-8166138
-      // c("\"1970-01-01T00:00:00+01:00\"", VA.timestamp)(Time.Timestamp assertFromLong 3600),
+      c("\"1970-01-01T00:00:00+01:00\"", VA.timestamp)(Time.Timestamp assertFromLong -3600000000L),
```
[the remark this answers is not in the document]
Comment (carlpulley-da) [reply sdk/canton/community/ledger/ledger-json-api/src/test/scala/com/digitalasset/canton/daml/lf/value/json/ApiCodecCompressedSpec.scala:331]: Thanks for that @tudor-da! 

Trick here is to initially use the string `1969-12-31T23:00:00Z` to compare against the `Timestamp.assertFromLong(-3600000000L)` value, then we can list `1970-01-01T00:00:00+01:00` as an alternative. Doing this allows this latter timestamp to get parsed and compared against this "base" value.

[the remark this answers is not in the document]
Comment (tudor-da) [reply sdk/canton/community/ledger/ledger-json-api/src/test/scala/com/digitalasset/canton/daml/lf/value/json/ApiCodecCompressedSpec.scala]: 👏

**Review by tudor-da — APPROVED** [review]
lgtm, thanks!

**Pull request merged**